### PR TITLE
Allow users to specify jackInEnv per replConnectSequences

### LINKED ISSUE #1124: Allow jackInEnv per replConnectSequences
Hello,

Thanks for all the work you all do on Calva, it's a great tool to make Clojure more accessible to the masses!

Currently in Calva, if one wished to send an `env` var to their project. There is `calva.jackInEnv` which lets one specify a JSON Object with environment variables for *all* connect sequences. I'd like to propose an option that lets one set these per object in `calva.replConnectSequences`. I imagine a `jackInEnv` could be provided in there and the json objects could be merged in the style of `{...global, ...customConnectJackIn}` where the more specific entries override the global ones.

The use case that I have is that I have different REPL connect sequences for development vs. running unit tests (different lein profiles are used). Additionally. I have different env variables that need to be used. Typically, I go into the vscode `settings.json` and change them prior to deciding what connect sequence I will be running. I'd be nice to just set the specific overrides that I need for unit tests inside of it's REPL custom connect sequence.

### PR DESCRIPTION
**Pull request by rjerue**
## What has Changed?

- Makes it possible to add a jackInEnv per replConnectSequences

Fixes #1124

## My Calva PR Checklist

I have:

- [x] Read [How to Contribute](https://github.com/BetterThanTomorrow/calva/wiki/How-to-Contribute#before-sending-pull-requests).
- [x] Directed this pull request at the `dev` branch. (Or have specific reasons to target some other branch.)
- [x] Made sure I have changed the PR base branch, so that it is not `published`. (Sorry for the nagging.)
- [x] Updated the `[Unreleased]` entry in `CHANGELOG.md`, linking the issue(s) that the PR is addressing.
- [x] Figured if **anything** about the fix warrants tests on Mac/Linux/Windows/Remote/Whatever, and either tested it there if so, or mentioned it in the PR.
- [x] Added to or updated docs in this branch, if appropriate
- [x] Tested the VSIX built from the PR (so, after you've submitted the PR). You'll find the artifacts by clicking _Show all checks_ in the CI section of the PR page, and then _Details_ on the `ci/circleci: build` test. NB: *You need to sign in/up at Circle CI to find the _Artifacts_ tab.*
     - [x] Tested the particular change
     - [x] Figured if the change might have some side effects and tested those as well.
     - [x] Smoke tested the extension as such.
- [x] Referenced the issue I am fixing/addressing _in a commit message for the pull request_.
     - [x] If I am fixing the issue, I have used [GitHub's fixes/closes syntax](https://help.github.com/en/articles/closing-issues-using-keywords)
     - [x] ~If I am fixing just part of the issue, I have just referenced it w/o any of the "fixes” keywords.~
- [x] Created the issue I am fixing/addressing, if it was not present.

## The Calva Team PR Checklist:

Before merging we (at least one of us) have:

- [x] Made sure the PR is directed at the `dev` branch (unless reasons).
- [x] Figured if **anything** about the fix warrants tests on Mac/Linux/Windows/Remote/Whatever, and tested it there if so.
- [x] Read the source changes.
- [x] Given feedback and guidance on source changes, if needed. (Please consider noting extra nice stuff as well.)
- [ ] Tested the VSIX built from the PR (well, if this is a PR that changes the source code.)
     - [ ] Tested the particular change
     - [ ] Figured if the change might have some side effects and tested those as well.
     - [ ] Smoke tested the extension as such.
- [ ] If need be, had a chat within the team about particular changes.

Ping @pez, @bpringe

### CODE REVIEW

**Comment by rjerue** [comment]
I looked hard for some tests, unfortunately, I didn't see anything that would test this particular feature, or the `getJackInTerminalOptions` function :(

**Comment by rjerue** [comment]
The local checks I did, if there's any nuances about `jackInEnv`, I'd be happy to test more.

Made the following settings.json

```json
{
  "calva.jackInEnv": {
    "hello": "world"
  },
  "calva.replConnectSequences": [
    {
      "name": "Lein Fun",
      "projectType": "Leiningen",
      "cljsType": "none",
      "jackInEnv": {
        "type": "beautiful"
      }
    }
  ]
}
```

Ran this in the REPL:

```clojure
(println "hello"  (System/getenv "type") (System/getenv "hello"))
```

Out:

```
clj꞉clojure-fun.core꞉> 
hello beautiful world
nil
```

I also tried negative cases (like no jackInEnv global/in replConnectSequences.

**Comment by bpringe** [comment]
Thanks for this enhancement! I'll leave a final review to @PEZ.

**Comment by bpringe** [comment]
Sorry, did that by accident...

**Comment by bpringe** [comment]
The only thing I see that maybe should be changed is the wording in the description of the property. "Overwrite" sounds like more of a replacement than a merge, whereas really the project and global environment properties are merged, with the project properties having priority.

**Comment by PEZ** [comment]
> I looked hard for some tests, unfortunately, I didn't see anything that would test this particular feature, or the getJackInTerminalOptions function :(

Indeed. There are zero tests that cover this, and that is something we really should try to fix.

File: docs/site/connect-sequences.md [hunk 1]
```diff
@@ -32,6 +32,7 @@ A connect sequence configures the following:
     * `cljAliases`: At Jack-in to a deps.edn project, use these aliases to launch the repl.
     * `cljsLaunchBuilds`: The cljs builds to start/watch at Jack-in/connect.
     * `cljsDefaultBuild`: Which cljs build to attach to at the initial connect.
+* `jackInEnv`: An object with environment variables that will be added to the environment of the Jack-in process. Will overwrite keys set at the global `calva.jackInEnv`.
```
Comment (PEZ) [inline docs/site/connect-sequences.md]:
```suggestion
* `jackInEnv`: An object with environment variables that will be merged with the global `calva.jackInEnv` and then applied to the Jack-in process. The merge is very similar to how Clojure's `merge` works. So for any common keys between the global setting and this one, the ones from this setting will win.
```

File: package.json [hunk 2]
```diff
@@ -489,6 +489,11 @@
                                             }
                                         }
                                     ]
+                                },
+                                "jackInEnv": {
+                                    "type": "object",
+                                    "default": {},
+                                    "description": "Specifies any environment variables for the given connect sequence. Will overwrite \"calva.jackInEnv\" keys of the same name"
```
Comment (PEZ) [inline package.json]:
```suggestion
                                    "description": "Environment variables that will be merged with any global `calva.jackInEnv` and then applied to the Jack-in process. The merge is very similar to how Clojure's `merge` works. So for any common keys between the global setting and this one, the ones from this setting will win."
```

**Comment by PEZ** [comment]
Yes, many thanks for this. It will be a very nice enhancement!

I agree with @bpringe 's observation about the wording. I suggested some alternatives. Use the suggestions as is or use them as inspiration. 😄 

Also think that the local env setting might be missing some of the treatment we currently give the global settings. Not entirely sure about that, and am a bit too tired right now to be able to figure it out, but please check and see if my suspicion is right or wrong.

**Comment by bpringe** [comment]
Sorry, I had deleted my comment I posted about the wording :smile:. I saw that he mentioned the "keys of the same name" being replaced and figured it was okay, but a different wording would probably be good, indeed.

**Comment by rjerue** [comment]
I'm good with the wording suggestions you made! @PEZ

**Comment by rjerue** [comment]
I see some additional processing being done in that `resolveEnvVariables` function. I'm out of town until Sunday, but will update to do that same processing to the new object too.👍

**Comment by rjerue** [comment]
@PEZ both the global and and per connect sequence env objects should have the same processing now :)

File: package.json [hunk 3]
```diff
@@ -489,6 +489,11 @@
                                             }
                                         }
                                     ]
+                                },
+                                "jackInEnv": {
+                                    "type": "object",
+                                    "default": {},
+                                    "description": "Environment variables that will be merged with any global `calva.jackInEnv` and then applied to the Jack-in process. The merge is very similar to how Clojure's `merge` works. So for any common keys between the global setting and this one, the ones from this setting will win."
```
Comment (PEZ) [inline package.json]:
```suggestion
                                    "markdownDescription": "Environment variables that will be merged with any global `calva.jackInEnv` and then applied to the Jack-in process. The merge is very similar to how Clojure's `merge` works. So for any common keys between the global setting and this one, the ones from this setting will win."
```

**Comment by PEZ** [comment]
Completely for dropping this ball. Now had a look again. Everythings looks fine. I think maybe you need to merge latest dev again, because the changelog looked a bit oldish.

**Review by PEZ — APPROVED** [review]
Super duper stuff!

Maybe needs merging of dev, and smoke tested after that.

**Comment by rjerue** [comment]
FYI I've been using this at work this week on an OSX machine and it's been working well.

**Comment by PEZ** [comment]
Tons of thanks for this contribution @rjerue! It was a pleasure reviewing.

**Pull request merged**